python/felis/db/database_context.py

```diff
@@ -0,0 +1,859 @@
+        engine_dialect = _dialect_name(engine.url)
+        expected_dialect = self.DIALECT.value.lower()
+
+        if engine_dialect != expected_dialect:
+            raise DatabaseContextError(
+                f"Engine dialect '{engine_dialect}' does not match the context's expected dialect: "
+                f"{expected_dialect}"
+            )
+
+    @property
+    def engine(self) -> Engine:
+        return self._engine
+
+    @property
+    def metadata(self) -> MetaData:
+        return self._metadata
+
+    @property
+    def dialect(self) -> Dialect:
+        return self._dialect
+
+    @property
+    def dialect_name(self) -> str:
+        """Get the dialect name for this database context.
+
+        Returns
+        -------
+        str
+            The normalized dialect name.
+        """
+        return self.DIALECT.value
+
+    @property
+    def schema_name(self) -> str | None:
+        """Effective schema name for this context (may be None).
+
+        Returns
+        -------
+        str | None
+            The schema name, or None if no schema is set.
+        """
+        return self._schema_name
+
+    def execute(self, statement: SQLStatement) -> Result:
+        statement = _normalize_statement(statement)
+        try:
+            with self.engine.begin() as connection:
+                result = connection.execute(statement)
+                return result
+        except SQLAlchemyError as e:
+            raise DatabaseContextError("Error executing statement", e)
+
+    def create_all(self) -> None:
+        with self.engine.begin() as conn:
+            try:
+                self.metadata.create_all(bind=conn)
+            except SQLAlchemyError as e:
+                raise DatabaseContextError("Error creating database", e)
+
+    def create_indexes(self) -> None:
+        with self.engine.begin() as conn:
+            try:
+                inspector = inspect(conn)
+                for table in self.metadata.tables.values():
+                    # Fetch all existing indexes for this table once
+                    existing_indexes = _get_existing_indexes(inspector, table.name, self.schema_name)
+
+                    for index in table.indexes:
+                        if index.name is None:
+                            # Anonymous indexes can't be checked by name
+                            logger.warning(f"Skipping anonymous index on table '{table.name}'")
+                            continue
+
+                        if index.name in existing_indexes:
+                            logger.warning(f"Skipping creation of index '{index.name}' which already exists")
+                            continue
+
+                        index.create(bind=conn, checkfirst=False)  # We already checked
+                        logger.info(f"Created index '{index.name}'")
+            except SQLAlchemyError as e:
+                raise DatabaseContextError("Error creating indexes", e)
+
+    def drop_indexes(self) -> None:
```

Comment:
The create/drop indexes methods seem mostly identical, suggestion to extract into a common method, something like `def _manage_index(self, action: str)`
and then they become:

```
def create_indexes(self) -> None:
    self._manage_index("create")

def drop_indexes(self) -> None:
    self._manage_index("drop")
```